crates/wasmtime/src/compiler.rs

```diff
@@ -0,0 +1,700 @@
+        translation: &'a ModuleTranslation<'a>,
+        functions: PrimaryMap<DefinedFuncIndex, FunctionBodyData<'a>>,
+    ) -> Self {
+        let mut inputs = vec![];
+        let module_index = StaticModuleIndex::from_u32(0);
+
+        Self::collect_inputs_in_translations(
+            types,
+            std::iter::once((module_index, translation, functions)),
+            &mut inputs,
+        );
+
+        CompileInputs { inputs }
+    }
+
+    /// Create a `CompileInputs` for a component.
+    #[cfg(feature = "component-model")]
+    pub fn for_component(
+        types: &'a wasmtime_environ::component::ComponentTypes,
+        component: &'a wasmtime_environ::component::Component,
+        module_translations: impl IntoIterator<
+            Item = (
+                StaticModuleIndex,
+                &'a ModuleTranslation<'a>,
+                PrimaryMap<DefinedFuncIndex, FunctionBodyData<'a>>,
+            ),
+        >,
+    ) -> Self {
+        let mut inputs = vec![];
+
+        Self::collect_inputs_in_translations(
+            types.module_types(),
+            module_translations,
+            &mut inputs,
+        );
+
+        for init in &component.initializers {
+            match init {
+                wasmtime_environ::component::GlobalInitializer::AlwaysTrap(always_trap) => {
+                    push_input(&mut inputs, move |_tunables, compiler| {
+                        Ok(CompileOutput {
+                            key: CompileKey::always_trap(always_trap.index),
+                            symbol: always_trap.symbol_name(),
+                            function: compiler
+                                .component_compiler()
+                                .compile_always_trap(&types[always_trap.canonical_abi])?
+                                .into(),
+                            info: None,
+                        })
+                    });
+                }
+                wasmtime_environ::component::GlobalInitializer::Transcoder(transcoder) => {
+                    push_input(&mut inputs, move |_tunables, compiler| {
+                        Ok(CompileOutput {
+                            key: CompileKey::transcoder(transcoder.index),
+                            symbol: transcoder.symbol_name(),
+                            function: compiler
+                                .component_compiler()
+                                .compile_transcoder(component, transcoder, types)?
+                                .into(),
+                            info: None,
+                        })
+                    });
+                }
+                wasmtime_environ::component::GlobalInitializer::LowerImport(lower_import) => {
+                    push_input(&mut inputs, move |_tunables, compiler| {
+                        Ok(CompileOutput {
+                            key: CompileKey::lowering(lower_import.index),
+                            symbol: lower_import.symbol_name(),
```

Comment:
These `symbol_name` functions may be best to move here now since they should only be called from this location?

Reply:
Maybe? I kind of like them as methods on the thing.